driver: display: add display support for the rt1170

CristXu opened the pull request:
Add the display support for the mimxrt1170 evk, including
the mipi-dsi, video mux, rm68200 lcd panel.

Signed-off-by: Crist Xu <crist.xu@nxp.com>

gmarull committed:
drivers: mipi_dsi: initial support for MIPI-DSI drivers

Add initial support for MIPI-DSI drivers.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

jfischer-no commented on drivers/display/display_rm68200.c:
```diff
@@ -0,0 +1,455 @@
+	{0x11, 0x00},
+};
+
+extern int mipi_dsi_dcs_write(const struct device *dev, const uint8_t *buf, uint8_t len);
```
This looks like the approach "My board/display works, rest does not matter". I think there is no way around mipi-dsi driver API.

CristXu replied:
Ok， I will fix this by open a repo for the rm6800 panel soon, with dts including the configuration, and get the info from this.

CristXu commented on drivers/display/display_mcux_elcdif.c:
```diff
@@ -252,6 +252,26 @@ static struct mcux_elcdif_config mcux_elcdif_config_1 = {
 	.pixel_format = PIXEL_FORMAT_BGR_565,
 	.bits_per_pixel = 16,
 #endif
+#ifdef CONFIG_MCUX_ELCDIF_PANEL_RM68200
```
Ok, Will fix it.

mmahadevan108 replied:
The display details have been moved to the device tree as part of PR https://github.com/zephyrproject-rtos/zephyr/pull/42429

mmahadevan108 replied:
Done

CristXu commented on drivers/display/display_mcux_elcdif.c:
```diff
@@ -198,7 +198,7 @@ static int mcux_elcdif_init(const struct device *dev)
 			return -ENOMEM;
 		}
-		memset(data->fb[i].data, 0, data->fb_bytes);
+		memset(data->fb[i].data, 0x56, data->fb_bytes);
```
Deleted

mmahadevan108 replied:
This change has been deleted.

gmarull reviewed (changes requested):
MIPI-DSI should be implemented as a vendor independent API, what is proposed here is effectively display drivers that are vendor dependent (NXP), which is wrong. In case you are interested, I started writing a MIPI-DSI API a long time ago (implemented for STM32) but never got finished (see [here](https://github.com/teslabs/zephyr/tree/feature/stm32-dsi)). I'm willing to help on making a proposal.

dleach02 commented:
> MIPI-DSI should be implemented as a vendor independent API, what is proposed here is effectively display drivers that are vendor dependent (NXP), which is wrong. In case you are interested, I started writing a MIPI-DSI API a long time ago (implemented for STM32) but never got finished (see [here](https://github.com/teslabs/zephyr/tree/feature/stm32-dsi?rgh-link-date=2022-01-24T18%3A14%3A29Z)). I'm willing to help on making a proposal.

@gmarull thank you for the reference. The recommendation makes sense and we are investigating it now.

mmahadevan108 commented on drivers/display/Kconfig.mcux_elcdif:
```diff
@@ -23,7 +23,7 @@ config MCUX_ELCDIF_POOL_BLOCK_MIN
 config MCUX_ELCDIF_POOL_BLOCK_MAX
 	hex "Maximum memory pool block size"
-	default 0x40000
+	default 0x1D0000
```
This is due to the increase display buffer size requirement as the display on the RT1170 is 720x1280. This change has been moved to the `boards` folder

mmahadevan108 commented on drivers/display/display_rm68200.c:
```diff
@@ -0,0 +1,455 @@
+	gpio_dt_flags_t bl_flags;
+};
+
+static const uint8_t rm68200InitParam[][2] = {
```
This driver is for the RM68200 display chip and these settings are specific to this chip.

mbolivar-nordic commented:
> I started writing a MIPI-DSI API a long time ago (implemented for STM32) but never got finished (see [here](https://github.com/teslabs/zephyr/tree/feature/stm32-dsi)). I'm willing to help on making a proposal.

Wow, a lot of work in here.

Looks like there are pretty well-developed Linux bindings for this IP: https://github.com/torvalds/linux/blob/master/Documentation/devicetree/bindings/display/st%2Cstm32-dsi.yaml

@gmarull your bindings look different; is that on purpose?

gmarull commented:
> > I started writing a MIPI-DSI API a long time ago (implemented for STM32) but never got finished (see [here](https://github.com/teslabs/zephyr/tree/feature/stm32-dsi)). I'm willing to help on making a proposal.
> 
> Wow, a lot of work in here.
> 
> Looks like there are pretty well-developed Linux bindings for this IP: [torvalds/linux@`master`/Documentation/devicetree/bindings/display/st%2Cstm32-dsi.yaml](https://github.com/torvalds/linux/blob/master/Documentation/devicetree/bindings/display/st%2Cstm32-dsi.yaml)
> 
> @gmarull your bindings look different; is that on purpose?

It was just an experiment, really. Afaik, NXP and STM32 use the same MIPI-DSI IP (Synopsys), so maybe a common driver could be made for both platforms (similar case: https://github.com/zephyrproject-rtos/zephyr/pull/39167).

mmahadevan108 committed:
modules: NXP: Add MIPI DSI support

Add support for the MXRT MIPI DSI controller

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
drivers: display: Add support for Raydium RM68200

Add support for the Raydium RM68200 MIPI DSI panel.

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
soc: rt: Add MIPI support for MXRT1170

Add MIPI support for NXP's MXRT1170

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
boards: Add MIPI support on MXRT1170 EVK

1. Add MIPI support on MXRT1170 EVK board

2. To support the 720x1280 on RT1170 EVK, we
   require more heap for the display buffers.
   Increase the heap size via HEAP_MEM_POOL_SIZE config

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
display: Update ELCDIF driver to use a config

Update the MCUX ELCDIF driver to use CONFIG_MCUX_ELCDIF_POOL_BLOCK_NUM
for the number for frame buffers to allocate.

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
samples: display: Increase Heap size for RT1170 EVK

Increase the heap size for RT1170 EVK to allocate the
display buffers for a 720x1280 display

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 commented on drivers/display/display_mcux_mipi_dsi.c:
```diff
@@ -0,0 +1,224 @@
+/*
```
The drivers have been separated into individual commits.

mmahadevan108 commented on boards/arm/mimxrt1170_evk/mimxrt1170_evk_cm7.dts:
```diff
@@ -26,6 +26,25 @@
 	};
 };
 
+&lcdif {
```
Done

mmahadevan108 commented:
> MIPI-DSI should be implemented as a vendor independent API, what is proposed here is effectively display drivers that are vendor dependent (NXP), which is wrong. In case you are interested, I started writing a MIPI-DSI API a long time ago (implemented for STM32) but never got finished (see [here](https://github.com/teslabs/zephyr/tree/feature/stm32-dsi)). I'm willing to help on making a proposal.

@gmarull I have updated this PR to use the experimental API that you have proposed. Thank you. It would be great if we can discuss next steps to get this into the mainline.

CristXu commented on soc/arm/nxp_imx/rt/soc_rt11xx.c:
```diff
@@ -520,6 +526,63 @@ void imxrt_audio_codec_pll_init(uint32_t clock_name, uint32_t clk_src,
+{
+	/* elcdif output to MIPI DSI */
+	CLOCK_EnableClock(kCLOCK_Video_Mux);
+	VIDEO_MUX->VID_MUX_CTRL.CLR = VIDEO_MUX_VID_MUX_CTRL_MIPI_DSI_SEL_MASK;
```
Here about the VIDEO_MUX, do we need to place it in another commit with a dts  defined?

mmahadevan108 replied:
If there is a need for a video mux driver. At this point we do not need one.

jfischer-no commented on drivers/display/Kconfig.rm68200:
```diff
@@ -0,0 +1,8 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config RM68200
+	bool "RM68200 display driver"
```
RM68200 is a MIPI DSI panel driver, that should be mentioned in the commit message as well.

mmahadevan108 replied:
Done, I have updated the commit message

jfischer-no commented on samples/drivers/display/boards/mimxrt1170_evk_cm7.conf:
```diff
@@ -0,0 +1,8 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+
+CONFIG_HEAP_MEM_POOL_SIZE=4194304
```
Taking in the account previous commit, why are you adding this overlay?

mmahadevan108 replied:
This is needed as we require more memory to allocate the frame buffers allocated in `driver_mcux_elcdif.c`

mmahadevan108 replied:
This is because `CONFIG_HEAP_MEM_POOL_SIZE=16384` is defined in `samples/drivers/display/prj.conf`. This is eliminating the default value defined in the boards `Kconfig.defconfig` file. I would prefer not to have this in the overlay but I did not want to impact other platforms display testing.

jfischer-no replied:
Right. Do you need to set it in [boards/arm/mimxrt1170_evk/Kconfig.defconfig](https://github.com/zephyrproject-rtos/zephyr/pull/36875/commits/57132abf74aa5d66b97c7c794b547727c35f4912#diff-03a9c4ae2d435ebefab2f185a68774643ae7cfd58b36a3a62cf9e4473dc33470) then? Because as I see it LVGL sample does not need it.

mmahadevan108 replied:
This board needs the extra `HEAP`, so I would like to keep it in the board file so that application developers do not have to worry about this.

gmarull replied:
it looks like the setting is adjusted in two places, defconfig and here. why isn't defconfig sufficient?

mmahadevan108 replied:
The tests `proj.conf` file overrides the defconfig value. `samples/drivers/display/prj.conf`

gmarull replied:
boards/arm/mimxrt1170_evk/Kconfig.defconfig:

```
# Memory from the heap pool is used to allocate display buffers.
# Adjust HEAP_MEM_POOL_SIZE.
config HEAP_MEM_POOL_SIZE
	default 4194304
```

samples/drivers/display/boards/mimxrt1170_evk_cm7.conf
```
CONFIG_HEAP_MEM_POOL_SIZE=4194304
```

ie both set `4194304`

mmahadevan108 replied:
`samples/drivers/display/prj.conf` overwrites the value set inside `boards/arm/mimxrt1170_evk/Kconfig.defconfig`. Hence this was added in the `samples/drivers/display/boards/mimxrt1170_evk_cm7.conf` file.

jfischer-no commented on drivers/display/Kconfig.rm68200:
```diff
@@ -3,6 +3,6 @@
 config RM68200
 	bool "RM68200 display driver"
-	depends on DISPLAY_MCUX_MIPI_DSI
+	depends on MIPI_DSI
```
There should be just one commit that introduces rm68200 panel driver instead of two.

mmahadevan108 replied:
Apologies for overloading this PR. This will get fixed once we have a MIPI API PR finalized and merged.

jfischer-no commented on drivers/display/Kconfig.mcux_mipi_dsi:
```diff
@@ -0,0 +1,8 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config DISPLAY_MCUX_MIPI_DSI
+	bool "NXP MCUX MIPI DSI driver"
```
Why this commit introduces a shim driver instead of a native one for a "well-known" DSI IP?
https://github.com/zephyrproject-rtos/zephyr/pull/36875#issuecomment-1047629683

mmahadevan108 replied:
This was added as a first step to prove the functionality. It was deleted in the later commit when we use the MIPI experimental API.

mmahadevan108 committed:
drivers: mipi: Add MIPI DSI driver for MXRT devices

Add a MIPI DSI driver for the MIPI controller on
NXP MXRT SoC'sUpdate MIPI_DSI

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 commented:
I have updated this PR to cleanup and squash some of the commits. This PR is using the experimental MIPI API proposed by @gmarull.
@carlescufi We would like to bring this to the zephyr API meeting for review.

mbolivar-nordic commented on dts/bindings/mipi-dsi/mipi-dsi-device.yaml:
```diff
@@ -0,0 +1,27 @@
+    label:
+      required: true
+
+    data-lanes:
+      type: int
+      required: true
```
I don't know anything about DSI, but the Linux bindings seem to use an array for this property; why is this an int?

mmahadevan108 replied:
@mbolivar-nordic , can you please point me to the Linux bindings file.

mbolivar-nordic replied:
I'm just grepping around in linux/Documentation/dts/bindings/display. Here are a few places where data-lanes looks like it might be being used as an array of lanes rather than a number of lanes:

```
$ rg data-lanes
msm/dsi-controller-main.yaml
101:              data-lanes:
117:              data-lanes:
199:                                   data-lanes = <0 1 2 3>;

tegra/nvidia,tegra20-host1x.txt
129:	 - data-lanes: an array of data lane from 1 to 8. Valid array
530:								data-lanes = <1 2>;

bridge/sii902x.txt
18:	- sil,i2s-data-lanes: Array of up to 4 integers with values of 0-3
63:		sil,i2s-data-lanes = < 0 1 2 >;

bridge/renesas,dsi-csi2-tx.yaml
62:              data-lanes:
67:              - data-lanes
112:                    data-lanes = <1 2>;

bridge/ti,sn65dsi86.yaml
97:              data-lanes:
147:              lane-polarities: [data-lanes]
270:              data-lanes = <2 1 3 0>;

bridge/lontium,lt8912b.yaml
43:              data-lanes: true
46:              - data-lanes
87:              data-lanes = <0 1 2 3>;

bridge/toshiba,tc358775.yaml
135:                    data-lanes = <0 1 2 3>;
201:                    data-lanes = <0 1 2 3>;

bridge/analogix,anx7625.yaml
109:              data-lanes: true
159:                        data-lanes = <0 1 2 3>;

bridge/ti,sn65dsi83.yaml
50:              data-lanes:
70:              data-lanes:
146:                        data-lanes = <1 2 3 4>;
```

I'm not sure if this is applicable to your binding, however, since I see this binding is about the device side, not the host side. Just double checking with you on this since as I said I don't really know anything about the DSI protocol.

mmahadevan108 replied:
@gmarull Can you help comment on this feedback from @mbolivar

gmarull replied:
the property describes the number of data lanes available on the bus (up to 4 in MIPI-DSI).

What Linux does:

```
  data-lanes:
    $ref: /schemas/types.yaml#/definitions/uint32-array
    minItems: 1
    maxItems: 8
    items:
      # Assume up to 9 physical lane indices
      maximum: 8
    description:
      An array of physical data lane indexes. Position of an entry determines
      the logical lane number, while the value of an entry indicates physical
      lane, e.g. for 2-lane MIPI CSI-2 bus we could have "data-lanes = <1 2>;",
      assuming the clock lane is on hardware lane 0. If the hardware does not
      support lane reordering, monotonically incremented values shall be used
      from 0 or 1 onwards, depending on whether or not there is also a clock
      lane. This property is valid for serial busses only (e.g. MIPI CSI-2).
```

I don't think we need that description now, but Linux probably has chosen it for reasons. Considering Zephyr video/display support is in its infancy stage, I think we should follow Linux.

mmahadevan108 replied:
Sorry, I was not clear on your earlier comment. I have updated to convert this to a array.

mbolivar-nordic commented on dts/bindings/display/nxp,imx-mipi-dsi.yaml:
```diff
@@ -0,0 +1,66 @@
+        Behavior in BLLP (Blanking or Low-Power Interval).
+
+    autoinsert-eotp:
+      type: boolean
+      required: true
```
`type: boolean` with `required: true` is a little weird... are you sure? I think this should be dropped.

mmahadevan108 replied:
Thank you @mbolivar-nordic. I have removed the line and have also moved the bindings file to the new `mipi-dsi` folder instead of `display`

mbolivar-nordic reviewed (commented):
Looks very nice! A couple of minor things on the bindings.

mbolivar-nordic commented on drivers/display/display_rm68200.c:
```diff
@@ -0,0 +1,1430 @@
+#define RM68200_PANEL(id)							\
+	static const struct rm68200_config rm68200_config_##id = {		\
+		.mipi_dsi_label = DT_INST_BUS_LABEL(id),			\
+		.reset_label = UTIL_AND(DT_INST_NODE_HAS_PROP(id, reset_gpios),	\
+					DT_INST_GPIO_LABEL(id, reset_gpios)),	\
+		.reset_pin = UTIL_AND(DT_INST_NODE_HAS_PROP(id, reset_gpios),	\
+				      DT_INST_GPIO_PIN(id, reset_gpios)),	\
+		.reset_flags = UTIL_AND(DT_INST_NODE_HAS_PROP(id, reset_gpios),	\
+					DT_INST_GPIO_FLAGS(id, reset_gpios)),	\
+		.pwr_label = UTIL_AND(DT_INST_NODE_HAS_PROP(id, pwr_gpios),	\
+				      DT_INST_GPIO_LABEL(id, pwr_gpios)),	\
+		.pwr_pin = UTIL_AND(DT_INST_NODE_HAS_PROP(id, pwr_gpios),	\
+				    DT_INST_GPIO_PIN(id, pwr_gpios)),		\
+		.pwr_flags = UTIL_AND(DT_INST_NODE_HAS_PROP(id, pwr_gpios),	\
+				      DT_INST_GPIO_FLAGS(id, pwr_gpios)),	\
+		.bl_label = UTIL_AND(DT_INST_NODE_HAS_PROP(id, bl_gpios),	\
+				     DT_INST_GPIO_LABEL(id, bl_gpios)),		\
+		.bl_pin = UTIL_AND(DT_INST_NODE_HAS_PROP(id, bl_gpios),		\
+				   DT_INST_GPIO_PIN(id, bl_gpios)),		\
+		.bl_flags = UTIL_AND(DT_INST_NODE_HAS_PROP(id, bl_gpios),	\
+				     DT_INST_GPIO_FLAGS(id, bl_gpios)),		\
```
I was taking a look at the driver since I was already here for the bindings and noticed this. It should be using GPIO_DT_SPEC_GET_OR() in all cases.

mmahadevan108 replied:
Thanks for your review, I have update the PR.

mbolivar-nordic commented on drivers/display/display_rm68200.c:
```diff
@@ -0,0 +1,1430 @@
+
+#define RM68200_PANEL(id)							\
+	static const struct rm68200_config rm68200_config_##id = {		\
+		.mipi_dsi_label = DT_INST_BUS_LABEL(id),			\
```
Why save a label when you can get the bus device pointer instead?

mmahadevan108 replied:
Thanks for your review, I have update the PR.

mmahadevan108 committed:
doc: Add MIPI API as an Experimental API

Add support for MIPI API

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 commented:
Rebased and updated this branch to include documentation in the `Experimental API` section and updated the `3.1 release notes`. I am working to add some API tests.

mmahadevan108 committed:
tests: Add MIPI API tests

Unit tests to test the MIPI API

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

gmarull commented on drivers/mipi_dsi/Kconfig:
```diff
@@ -0,0 +1,23 @@
+# Copyright (c) 2020 Teslabs Engineering S.L.
+# SPDX-License-Identifier: Apache-2.0
+
+menuconfig MIPI_DSI
+	bool "MIPI-DSI Host Controllers [EXPERIMENTAL]"
+	help
+	  Add support for MIPI-DSI host controllers
```
nit: `select EXPERIMENTAL`

mmahadevan108 replied:
Done

gmarull commented on drivers/mipi_dsi/mipi_dsi.c:
```diff
@@ -0,0 +1,106 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <drivers/mipi_dsi.h>
```
nit: use `<zephyr/...>`

mmahadevan108 replied:
Done

gmarull commented on include/zephyr/drivers/mipi_dsi.h:
```diff
@@ -0,0 +1,352 @@
+				  uint8_t channel,
+				  const struct mipi_dsi_device *mdev)
+{
+	const struct mipi_dsi_driver_api *api = dev->api;
```
nit: missing cast for C++ support

mmahadevan108 replied:
Done

gmarull commented on include/zephyr/dt-bindings/mipi_dsi/mipi_dsi.h:
```diff
@@ -0,0 +1,23 @@
+#define ZEPHYR_INCLUDE_DT_BINDINGS_MIPI_DSI_MIPI_DSI_H_
+
+/**
+ * @name MIPI-DSI Pixel formats.
+ */
+
+/** RGB888 (24bpp). */
+#define MIPI_DSI_PIXFMT_RGB888		0U
+/** RGB666 (24bpp). */
+#define MIPI_DSI_PIXFMT_RGB666		1U
+/** Packed RGB666 (18bpp). */
+#define MIPI_DSI_PIXFMT_RGB666_PACKED	2U
+/** RGB565 (16bpp). */
+#define MIPI_DSI_PIXFMT_RGB565		3U
```
nit: missing `@{` and `@}` (after `@name`)

gmarull commented on doc/hardware/peripherals/mipi.rst:
```diff
@@ -0,0 +1,13 @@
+MIPI
+####
+
+Overview
+********
+
```
if overview is empty, just delete it or add some introduction.

mmahadevan108 replied:
Deleted

gmarull commented on doc/hardware/peripherals/mipi.rst:
```diff
@@ -0,0 +1,13 @@
+.. _mipi_api:
+
+MIPI
```
`MIPI-DSI` (MIPI has other standards, e.g. MIPI-CSI2). same for file name (`mipi_dsi.rst`)

mmahadevan108 replied:
Renamed to `mipi_dsi`

gmarull commented on doc/hardware/peripherals/mipi.rst:
```diff
@@ -0,0 +1,13 @@
+API Reference
+*************
+
+.. doxygengroup:: mipi_interface
```
doxygen group should be called `mipi_dsi_interface`

mmahadevan108 replied:
Done

gmarull commented on doc/releases/release-notes-3.1.rst:
```diff
@@ -74,6 +74,10 @@ New APIs in this release
 
+* MIPI
+
+  * Added a new MIPI API for MIPI-DSI drivers.
```
mention that it is experimental (as it has some features/API calls that will likely be implemented later)

gmarull commented on modules/Kconfig.mcux:
```diff
@@ -63,6 +63,11 @@ config HAS_MCUX_ELCDIF
+config HAS_MCUX_MIPI_DSI
+	bool
+	help
+	  Set if the MIPI DSI moduel is present in the SoC.
```
`module`

mmahadevan108 replied:
Thank you. Fixed

gmarull commented on soc/arm/nxp_imx/rt/Kconfig.defconfig.series:
```diff
@@ -28,6 +28,10 @@ config DISPLAY_MCUX_ELCDIF
 	default y if HAS_MCUX_ELCDIF
 	depends on DISPLAY
 
+config MIPI_DSI_MCUX
+	default y if HAS_MCUX_MIPI_DSI
+	depends on MIPI_DSI
```
note: this should not be needed if the driver Kconfig defaults based on DT compatible status. You can find examples in a few areas, e.g. pinctrl drivers.

mmahadevan108 replied:
Thanks for pointing this, I have made the change

gmarull commented on soc/arm/nxp_imx/rt/soc_rt11xx.c:
```diff
@@ -492,6 +498,63 @@ void imxrt_audio_codec_pll_init(uint32_t clock_name, uint32_t clk_src,
 #endif
 
+#if CONFIG_MIPI_DSI
+void imxrt_pre_init_display_interface(void)
```
note: likely for future, but this is doing clock control work, so there should be a proper clock control driver handling that (unfortunately clock control area is a bit of a mess now)

mmahadevan108 replied:
Yes, we will work to improve this and other drivers once we have a more improved clocking infrastructure in place

gmarull commented on drivers/mipi_dsi/Kconfig.mcux:
```diff
@@ -0,0 +1,8 @@
+
+config MIPI_DSI_MCUX
+	bool "NXP MCUX MIPI-DSI Host Controller"
+	depends on HAS_MCUX_MIPI_DSI
```
see my comment about enabling based on dt compat status

mmahadevan108 replied:
Done.

gmarull commented on drivers/mipi_dsi/dsi_mcux.c:
```diff
@@ -0,0 +1,176 @@
+
+#define DT_DRV_COMPAT nxp_imx_mipi_dsi
+
+#include <drivers/video.h>
+#include <drivers/mipi_dsi.h>
+#include <fsl_mipi_dsi.h>
+#include <fsl_clock.h>
+#include <logging/log.h>
```
nit: please update using `<zephyr/>`

mmahadevan108 replied:
Done

gmarull commented on drivers/mipi_dsi/dsi_mcux.c:
```diff
@@ -0,0 +1,176 @@
+	dsi_config_t dsiConfig;
+	uint32_t mipiDsiEscClkFreq_Hz;
+	uint32_t mipiDsiTxEscClkFreq_Hz;
+	uint32_t mipiDsiDphyRefClkFreq_Hz = 24000000;
```
magic number here, why 24MHz?
another comment: please don't use `camelCase` naming in Zephyr code.

mmahadevan108 replied:
Thanks for catching. I moved this to device tree. Also I have fixed the `camelCase` naming of variables

gmarull commented on dts/bindings/mipi-dsi/nxp,imx-mipi-dsi.yaml:
```diff
@@ -0,0 +1,65 @@
+      description:
+        Instance of the LCDIF peripheral.
+
+    dpi-color-coding:
+      type: string
+      enum:
+        - "16-bit config 1"
+        - "16-bit config 2"
+        - "16-bit config 3"
+        - "18-bit config 1"
+        - "18-bit config 2"
+        - "24-bit"
+      description:
+        MIPI DPI interface color coding.
```
can "config N" meaning be described?

mmahadevan108 replied:
This is what I found documented in our reference manual:
`Sets the distribution of RGB bits within the 24-bit d bus, as specified by the DPI specification.`
I will add this to the property description.

gmarull commented on dts/bindings/mipi-dsi/nxp,imx-mipi-dsi.yaml:
```diff
@@ -0,0 +1,65 @@
+      description:
+        MIPI DSI pixel packet type send through DPI interface.
+
+    dpi-video-mode:
+      type: string
+      enum:
+        - "Non burst synch pulse"
+        - "Non burst synch event"
+        - "Burst"
+      description:
+        DPI video mode.
```
video mode is one of the MIPI-DSI modes (other is command, STM32 also has a sort of a hybrid), not sure I understand this property. is this an attribute for video mode?

another comment, string enums are `usually-written-like-this`, not `Written like this`.

mmahadevan108 replied:
Yes, this is a property for the video mode as expected by the Controller.
I have update the enums.

gmarull commented on dts/bindings/display/raydium,rm68200.yaml:
```diff
@@ -0,0 +1,48 @@
+        The BLn pin is asserted to control the backlight of the panel.
+        The sensor receives this as an active-high signal.
+
+    height:
+      type: int
+      required: true
+      description:
+        Height of the panel driven by the controller, with the units in pixels.
+
+    width:
+      type: int
+      required: true
+      description:
+        Width of the panel driven by the controller, with the units in pixels.
```
please make use of `display-controller.yaml`, these are already defined.

mmahadevan108 replied:
Thank you.

gmarull commented on dts/bindings/display/raydium,rm68200.yaml:
```diff
@@ -0,0 +1,48 @@
+        The RESETn pin is asserted to disable the sensor causing a hard
+        reset.  The sensor receives this as an active-low signal.
+
+    pwr-gpios:
+      type: phandle-array
+      required: false
+      description: |
+        The PWRn pin is asserted to do a power on for the panel.
+        The panel receives this as an active-high signal.
```
why not use the regulator API here? Define a regulator that handles display power, and, if displays needs to control power manually make use of it.

mmahadevan108 replied:
I don't much of a benefit in using regulator API here.

jfischer-no replied:
> I don't see much of a benefit in using regulator API here.

Please explain why not.

mmahadevan108 replied:
The use is fairly simple which is to enable a GPIO line. Hence my thought if there was a need to got through the regulator API.

gmarull commented on drivers/display/display_rm68200.c:
```diff
@@ -0,0 +1,1367 @@
+			return ret;
+		}
+
+		/* pwr:1 -> reset:0 -> reset:1 */
+		gpio_pin_set_dt(&config->pwr_gpio, 1);
+		k_sleep(K_MSEC(1));
+		gpio_pin_set_dt(&config->reset_gpio, 0);
+		k_sleep(K_MSEC(1));
+		gpio_pin_set_dt(&config->reset_gpio, 1);
+		k_sleep(K_MSEC(5));
```
magic sequence here: please document these timings

mmahadevan108 replied:
I have added some documentation from the datasheet. These were timings from our SDK example.

gmarull commented on drivers/display/display_rm68200.c:
```diff
@@ -0,0 +1,1367 @@
+	param[1] = DCS_CMD_PAGE_SET_PAGE2;
+	rm68200_generic_write(dev, param, 2);
+
+	/* There is no description for the below registers in the datasheet */
+	param[0] = 0x00;
+	param[1] = 0x05;
+	rm68200_generic_write(dev, param, 2);
+
+	param[0] = 0x02;
+	param[1] = 0x0B;
+	rm68200_generic_write(dev, param, 2);
+
+	param[0] = 0x03;
+	param[1] = 0x0F;
+	rm68200_generic_write(dev, param, 2);
+
+	param[0] = 0x04;
+	param[1] = 0x7D;
+	rm68200_generic_write(dev, param, 2);
```
it's a shame that vendor requires all these generic writes and doesn't document what they mean, maybe it's a kind of binary blob, who knows! Why not create a static const array holding all "magic numbers" and dump them in a loop, at least ugliness will be concentrated in a single array...

mmahadevan108 replied:
I have switched to using DCS writes. We had initially implemented using a static const array, later switched to this way per feedback received. This also matches the Linux driver implementation.

gmarull commented on drivers/display/display_rm68200.c:
```diff
@@ -0,0 +1,1367 @@
+	k_sleep(K_MSEC(200));
+
+	if (config->bl_gpio.port != NULL) {
+		ret = gpio_pin_configure_dt(&config->bl_gpio, GPIO_OUTPUT_INACTIVE);
+		if (ret < 0) {
+			LOG_ERR("Could not configure bl GPIO (%d)", ret);
+			return ret;
+		}
+
+		/* turn the backlight on */
+		ret = gpio_pin_set_dt(&config->bl_gpio, 1);
```
`GPIO_OUTPUT_ACTIVE` to skip the `pin_set_dt` call?

mmahadevan108 replied:
Thank you, I have updated.

gmarull commented on boards/arm/mimxrt1170_evk/Kconfig.defconfig:
```diff
@@ -52,4 +52,20 @@ config ETH_MCUX_PHY_RESET
 
+if MIPI_DSI
+
+config RM68200
+	default y
```
should not be required if display driver defaults to y based on DT compat being okay

mmahadevan108 replied:
Without this `CONFIG_RM68200` is not set

gmarull reviewed (changes requested):
Thanks for the contribution (and sorry for my late review). I have added a few comments/suggestions.

mmahadevan108 commented:
@gmarull @jfischer-no I have updated this PR to add a MIPI test case. I have covered all the comments mentioned in the API meeting when this API was first introduced. 
Can you please re-review and let me know if any additional changes are required.

gmarull commented on boards/arm/mimxrt1170_evk/mimxrt1170_evk_cm7.dts:
```diff
@@ -32,6 +33,46 @@
+		compatible = "raydium,rm68200";
+		reg = <0x0>;
+		label = "RM68200";
+		pwr-gpios = <&gpio11 16 GPIO_ACTIVE_HIGH>;
```
Linux driver seems to use the regulator framework, see https://elixir.bootlin.com/linux/v5.17.5/source/drivers/gpu/drm/panel/panel-raydium-rm68200.c#L379. I'd personally do the same here. However, if the driver doesn't support PM you may just define a regulator that is always on in DT, then display doesn't even need to take care of turning it on.

gmarull replied:
There are a few reasons. If the driver doesn't do any PM, the regulator can be configured to automatically be ON and so driver code doesn't even need to take care of managing display power. But it is not only that. Imagine there is a board where the display is part of a power rail that powers the display and something else. Then, the display can no longer manage the switch individually, so one would have to fork and modify the driver. This won't happen when using regulators, as they are reference counted.

mmahadevan108 replied:
I have switched to using the regulator API to be automatically ON.

mmahadevan108 commented on drivers/mipi_dsi/dsi_mcux.c:
```diff
@@ -0,0 +1,204 @@
+		dsi_xfer.dscCmd = msg->cmd;
+		dsi_xfer.txDataType = kDSI_TxDataDcsShortWrNoParam;
+		break;
+	case MIPI_DSI_DCS_SHORT_WRITE_PARAM:
+	case MIPI_DSI_DCS_LONG_WRITE:
```
Done

mmahadevan108 commented on drivers/mipi_dsi/dsi_mcux.c:
```diff
@@ -0,0 +1,204 @@
+	case MIPI_DSI_GENERIC_SHORT_WRITE_2_PARAM:
+		dsi_xfer.txDataType = kDSI_TxDataGenShortWrTwoParam;
+		break;
+	case MIPI_DSI_GENERIC_READ_REQUEST_0_PARAM:
+	case MIPI_DSI_GENERIC_READ_REQUEST_1_PARAM:
+	case MIPI_DSI_GENERIC_READ_REQUEST_2_PARAM:
```
Done

mmahadevan108 commented on include/zephyr/drivers/mipi_dsi.h:
```diff
@@ -0,0 +1,352 @@
+ */
+
+#include <stddef.h>
+#include <sys/types.h>
+
+#include <zephyr/types.h>
+#include <device.h>
+
+#include <dt-bindings/mipi_dsi/mipi_dsi.h>
```
Done

mmahadevan108 commented on tests/drivers/mipi_dsi/api/src/main.c:
```diff
@@ -0,0 +1,94 @@
+#define MIPI_NODE DT_INST(0, nxp_imx_mipi_dsi)
+#endif
+
+#ifdef MIPI_NODE
+#define MIPI_DEV_NAME DT_LABEL(MIPI_NODE)
+#else
```
Done

mmahadevan108 commented on drivers/mipi_dsi/dsi_mcux.c:
```diff
@@ -0,0 +1,204 @@
+			.videoMode = DT_INST_ENUM_IDX(id, dpi_video_mode),			\
+			.bllpMode = DT_INST_ENUM_IDX(id, dpi_bllp_mode),			\
+			.pixelPayloadSize = DT_PROP_BY_PHANDLE(DT_DRV_INST(id), nxp_lcdif, width),\
+			.panelHeight = DT_PROP_BY_PHANDLE(DT_DRV_INST(id), nxp_lcdif, height),	\
+			.polarityFlags = DT_PROP_BY_PHANDLE_IDX(				\
+					 DT_DRV_INST(id), nxp_lcdif, id, polarity) >> 2,	\
+			.hfp = DT_PROP_BY_PHANDLE(DT_DRV_INST(id), nxp_lcdif, hfp),		\
+			.hbp = DT_PROP_BY_PHANDLE(DT_DRV_INST(id), nxp_lcdif, hbp),		\
+			.hsw = DT_PROP_BY_PHANDLE(DT_DRV_INST(id), nxp_lcdif, hsync),		\
+			.vfp = DT_PROP_BY_PHANDLE(DT_DRV_INST(id), nxp_lcdif, vfp),		\
+			.vbp = DT_PROP_BY_PHANDLE(DT_DRV_INST(id), nxp_lcdif, vbp),		\
```
Thanks

mmahadevan108 commented on drivers/display/Kconfig.rm68200:
```diff
@@ -0,0 +1,8 @@
+# Copyright (c) 2022, NXP
+# SPDX-License-Identifier: Apache-2.0
+
+config RM68200
```
Updated.

dleach02 commented:
@gmarull @jfischer-no, we are pushing for getting this in the 3.1 feature freeze. Please review the feedback/changes @mmahadevan108 has provided to your comments to determine if your change requests have been satisfied.

gmarull reviewed (commented):
a few minor comments, otherwise lgtm

mmahadevan108 commented on tests/drivers/mipi_dsi/api/src/main.c:
```diff
@@ -0,0 +1,95 @@
+ * @}
+ */
+
+#if DT_HAS_COMPAT_STATUS_OKAY(nxp_imx_mipi_dsi)
+#define MIPI_NODE DT_INST(0, nxp_imx_mipi_dsi)
+#else
+#define MIPI_NODE ""
+#error "Unsupported SoC"
+#endif
+
+#ifndef CHANNEL
+#define CHANNEL 0
+#endif
```
This can be overwritten earlier when the platform specific `MIPI_NODE` is defined.

gmarull replied:
suggestion: create `mipi-dsi` alias, then in the test yaml: `filter: dt_alias_exists("mipi-dsi")`. Then test can just use the node defined by the alias.

mmahadevan108 replied:
I have made this change and get the channel number for a test `Kconfig`

mmahadevan108 commented on tests/drivers/mipi_dsi/api/src/main.c:
```diff
@@ -0,0 +1,95 @@
+	const struct device *dev = DEVICE_DT_GET(MIPI_NODE);
+	uint8_t rx_buf[2];
+	uint8_t param[2];
+	int ret;
```
Thank you

mmahadevan108 commented on tests/drivers/mipi_dsi/api/src/main.c:
```diff
@@ -0,0 +1,95 @@
+	memset(rx_buf, 0, sizeof(rx_buf));
+
+	param[0] = MIPI_DCS_GET_DISPLAY_BRIGHTNESS;
+	ret = mipi_dsi_generic_read(dev, CHANNEL, param, 1, rx_buf, 2);
+	if (ret != -ENOTSUP) {
+		zassert(ret >= 0, "Failed to read", NULL);
+	}
```
Thanks, done

mmahadevan108 commented on tests/drivers/mipi_dsi/api/src/main.c:
```diff
@@ -0,0 +1,95 @@
+	uint8_t param[2];
+	int ret;
+
+	zassert_not_null(dev, NULL);
```
Thanks

mmahadevan108 commented on tests/drivers/mipi_dsi/api/src/main.c:
```diff
@@ -0,0 +1,95 @@
+ */
+static void test_generic(void)
+{
+	const struct device *dev = DEVICE_DT_GET(MIPI_NODE);
```
Changed

mmahadevan108 commented on doc/hardware/peripherals/mipi_dsi.rst:
```diff
@@ -0,0 +1,11 @@
+
+MIPI-DSI
+########
+
+
+
```
Done

gmarull commented on tests/drivers/mipi_dsi/api/src/main.c:
```diff
@@ -0,0 +1,102 @@
+#define CHANNEL 0
+#endif
+
+static const struct device *mipi_dev;
```
```suggestion
static const struct device *mipi_dev = DEVICE_DT_GET(MIPI_NODE);
```

gmarull commented on tests/drivers/mipi_dsi/api/src/main.c:
```diff
@@ -0,0 +1,102 @@
+
+void test_main(void)
+{
+	mipi_dev = DEVICE_DT_GET(MIPI_NODE);
```
```suggestion
	__ASSERT_NO_MSG(device_is_ready(mipi_dev));
```

mmahadevan108 replied:
Done

gmarull replied:
not solved in https://github.com/zephyrproject-rtos/zephyr/commit/59e5d9a22ce2588166cae6a452649c2eeae57285

gmarull replied:
@mmahadevan108  forgot to push?

mmahadevan108 replied:
I have re-pushed, please check

mmahadevan108 commented:
@jfischer-no Could you take a look and let me know if all looks good per your review. 
@gmarull Can you please reapprove. I had to re-push to fix CI issues.
